### PR TITLE
Keycloak Authentication

### LINKED ISSUE #377: spark.jobserver.auth hasPermission
In the AuthInfo class (spark.jobserver.auth), the hasPermission method is empty. There is a comment that states "all authenticated users can perform all actions."

The AuthInfo class takes a "User" (or user name) as input.  Would it be good idea to create a pull request to add code to group users by roll and to grant permission by role?  What would be the best way to do this?

### PR DESCRIPTION
**Pull request by Ennosigaeon**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](doc/contribution-guidelines.md#commit-message-format) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

**New behavior :** 
This PR addresses three issues (each in a separate commit).
1. Refractoring of current Authenticator logic to use dedicated classes for each authentication method. This commit does not change existing behaviour at all.
2. Support for BasicAuth via [Keycloak](https://www.keycloak.org/) as a new Authentication Provider.
3. Introduction of a permission system to grant fine-granular access to single routes. This behaviour is currently only supported via Keycloak. For Shiro (and no authentication), users will automatically be assigned the `ALLOW_ALL` role which grants access to all routes. Resolves #377.

**BREAKING CHANGES** 

Previously, shiro configuration had to be provided via `shiro` prefix in the config file. The new prefix is `authentication.shiro` and config files have to be adapted accordingly.

**Other information**: 
I have split the changes in three commits. I assume, that commit 3 is the most controversial one. If desirable, I can move that commit to a dedicated PR.

---
This change is [Reviewable](https://reviewable.io/reviews/spark-jobserver/spark-jobserver/1338)

### CODE REVIEW

**Review by bsikander — COMMENTED** [review]
+1 for the change. Overall things look fine. I like that you have abstracted out the implementation of authentication.

A few questions:
- Can you give an example usecase where authorization would come in handy? I can see that with authorization, clients can only access endpoints that they are supposed to access and stop unwanted access but was there a specific requirement.
- You have defined the section `authentication` but with keycloak, it is also doing authorization. Have you considered naming it a bit different?

File: job-server/src/test/scala/spark/jobserver/auth/SJSAuthenticatorSpec.scala [hunk 1]
```diff
@@ -108,30 +56,250 @@ class SJSAuthenticatorSpec extends SJSAuthenticator with FunSpecLike
...
+
+      assert(instance.loginCount == 1)
+    }
+
+    it("should allow user with valid credentials after first rejection") {
+      val instance = new MockedKeycloakAuthenticator(config)
+      val cred = new BasicHttpCredentials(testUserWithValidGroup, testUserInvalidPassword)
+      Await.result(instance.challenge()(Some(cred)), 10.seconds) should equal(None)
+
+      val cred2 = new BasicHttpCredentials(testUserWithValidGroup, testUserWithoutValidGroupPassword)
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/test/scala/spark/jobserver/auth/SJSAuthenticatorSpec.scala:239]: Replaced `testUserWithoutValidGroupPassword` with `testUserWithValidGroupPassword`.

File: job-server/src/main/scala/spark/jobserver/auth/Permissions.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  val DATA_DELETE: Permission = Permission("data:delete", DATA)
+  val DATA_RESET: Permission = Permission("data:reset", DATA)
+
+  val JOBS: Permission = Permission("jobs")
+  val JOBS_READ: Permission = Permission("jobs:read", JOBS)
+  val JOBS_START: Permission = Permission("jobs:start", JOBS)
+  val JOBS_DELETE: Permission = Permission("jobs:delete", JOBS)
+
+  def apply(name: String): Option[Permission] = {
+    val permissions = Seq(ALLOW_ALL,
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/Permissions.scala:35]: Done

File: job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        case ex: JwkException =>
+          logger.warn(f"ACCESS DENIED (Invalid token), user [$user], ex [${ex.getMessage}]")
+          cache.remove(user)
+          None
+        case ex: Throwable =>
+          logger.info(f"ACCESS DENIED (Unexpected exception), user [$user], ex [${ex.getMessage}]")
+          cache.remove(user)
+          None
+      }
+    Await.result(f, authTimeout seconds)
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala:72]: Done

File: job-server/src/main/scala/spark/jobserver/auth/AllowAllAuthenticator.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package spark.jobserver.auth
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model.headers.BasicHttpCredentials
+import com.typesafe.config.Config
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+  * default authenticator that accepts all users
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/AllowAllAuthenticator.scala:10]: Done

File: doc/permissions.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+| DATA_READ       | data:read       | GET /data                                                                                                                                                    |
+| DATA_UPLOAD     | data:upload     | POST /data/<filename>                                                                                                                                        |
+| DATA_DELETE     | data:delete     | DELETE /data/<filename>                                                                                                                                      |
+| DATA_RESET      | data:reset      | PUT /data?reset=reboot                                                                                                                                       |
+| JOBS            | jobs            | GET /jobs <br /> GET /jobs/<jobId> <br />  GET /jobs/<jobId>/config <br /> DELETE /jobs/<jobId>                                                              |
+| JOBS_READ       | jobs:read       | GET /jobsGET /jobs/<jobId> <br /> GET /jobs/<jobId>/config                                                                                                   |
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply doc/permissions.md:24]: Done

[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply doc/permissions.md:24]: Fixed

File: README.md [hunk 6]
```diff
@@ -37,6 +37,7 @@ Also see [Chinese docs / 中文](doc/chinese/job-server.md).
     - [Server authentication](#server-authentication)
     - [Client authentication](#client-authentication)
   - [Basic authentication](#basic-authentication)
+  - [Permissions](#permissions)
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply README.md:40]: Reference to Keycloak and Shiro added. *Permissions* renamed to *User authorization* and readme rephrased.

File: job-server/src/test/resources/auth/shiro.ini [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+# search base for ldap groups:
+activeDirectoryRealm.contextFactory.environment[ldap.searchBase] = dc=xxx,dc=org
+activeDirectoryRealm.contextFactory.environment[ldap.allowedGroups] = "cn=xx,ou=groups", "cn=spark,ou=groups",,,,,
+activeDirectoryRealm.contextFactory.environment[java.naming.security.credentials] = password
+activeDirectoryRealm.contextFactory.url = ldap://localhost:389
+activeDirectoryRealm.userDnTemplate = cn={0},ou=people,dc=xxx,dc=org
+
+cacheManager = org.apache.shiro.cache.MemoryConstrainedCacheManager
+
+securityManager.cacheManager = $cacheManager
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/test/resources/auth/shiro.ini:14]: Done

File: job-server/src/test/resources/auth/dummy.ini [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+barf = 12345
+
+# -----------------------------------------------------------------------------
+# Roles with assigned permissions
+# roleName = perm1, perm2, ..., permN
+# -----------------------------------------------------------------------------
+[roles]
+admin = *
+schwartz = lightsaber:*
+goodguy = winnebago:drive:eagle5
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/test/resources/auth/dummy.ini:26]: Done

File: job-server/src/main/scala/spark/jobserver/auth/ShiroAuthenticator.scala [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.shiro.authz.AuthorizationException
+import org.apache.shiro.config.IniSecurityManagerFactory
+
+import scala.concurrent.ExecutionContext
+
+/**
+  * Apache Shiro based authenticator for the Spark JobServer, the authenticator realm must be
+  * specified in the ini file for Shiro
+  */
+class ShiroAuthenticator(override protected val config: Config)
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/ShiroAuthenticator.scala:18]: Renamed to `authConfig`.

File: job-server/src/main/scala/spark/jobserver/auth/SJSAuthenticator.scala [hunk 10]
```diff
@@ -1,110 +1,71 @@
...
+            logger.debug("No credentials provided")
+            None
         }
       }
       import akka.pattern.after
-      lazy val t = after(duration = authTimeout second,
+      lazy val t = after(duration = authTimeout seconds,
         using = s.scheduler)(Future.failed(new TimeoutException("Authentication timed out!")))
 
       Future firstCompletedOf Seq(f, t)
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/SJSAuthenticator.scala:64]: Done

File: job-server/src/main/scala/spark/jobserver/auth/SJSAuthenticator.scala [hunk 11]
```diff
@@ -1,110 +1,71 @@
...
+                None
+            }
+          case _ =>
+            logger.debug("No credentials provided")
+            None
         }
       }
       import akka.pattern.after
-      lazy val t = after(duration = authTimeout second,
+      lazy val t = after(duration = authTimeout seconds,
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/SJSAuthenticator.scala:61]: Done

File: job-server/src/main/scala/spark/jobserver/auth/ShiroAuthenticator.scala [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+      case ae: AuthorizationException =>
+        logger.info("ACCESS DENIED (" + ae.getMessage() + "), user [" + user + "]")
+        None
+      case ae: AuthenticationException =>
+        logger.info("ACCESS DENIED (Authentication Exception), user [" + user + "]")
+        None
+    }
+  }
+
+}
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/ShiroAuthenticator.scala:57]: Done

File: scalastyle-config.xml [hunk 13]
```diff
@@ -72,7 +72,7 @@
  <check level="error" class="org.scalastyle.scalariform.NoCloneChecker" enabled="true"></check>
  <check level="error" class="org.scalastyle.scalariform.NoFinalizeChecker" enabled="true"></check>
  <check level="error" class="org.scalastyle.scalariform.CovariantEqualsChecker" enabled="true"></check>
- <check level="error" class="org.scalastyle.scalariform.StructuralTypeChecker" enabled="true"></check>
+ <check level="error" class="org.scalastyle.scalariform.StructuralTypeChecker" enabled="false"></check>
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply scalastyle-config.xml:75]: I have to implement a `SigningKeyResolver` from `Jwts` in `KeycloakAuthenticator.` The interface requires methods using 
```public interface JwsHeader<T extends JwsHeader<T>> extends Header<T>```
This type annotation is quite ugly in scala
```
JwsHeader[T] forSome {type T <: JwsHeader[T]}
```
I am not sure how to implement a Scala type without using structural typing. I am happy to change the implementation in `KeycloakAuthenticator` if you have a better idea.

File: doc/permissions.md [hunk 14]
```diff
@@ -0,0 +1,29 @@
+# Permissions
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply doc/permissions.md:1]: I have not secured those endpoints on purpose. The `healthz` route should always be available to keep monitoring checks simple. UI is a bit more controversial: I think the UI itself does not expose sensitive data. Only information loaded in the UI (which are secured via authorization) are sensitive. I just noticed, that I did not check what happens if I open the UI and do not have, for example, permissions to read job data. I will have a look at that.

File: job-server/src/main/scala/spark/jobserver/auth/AllowAllAuthenticator.scala [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+      Future {
+        authenticate(BasicHttpCredentials.apply("", ""))
+      }
+    }
+  }
+
+  override def authenticate(credentials: BasicHttpCredentials): Option[AuthInfo] =
+    Some(new AuthInfo(User("anonymous")))
+
+}
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/AllowAllAuthenticator.scala:26]: Done

File: job-server/src/main/resources/application.conf [hunk 16]
```diff
@@ -319,17 +319,26 @@ akka.http.server {
   parsing.max-content-length = 30m
 }
 
-shiro {
-    # activate authentication (and authorization)
-    authentication = off
+authentication {
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/resources/application.conf:322]: If you let me know how you handle release notes, I can prepare that information.

File: doc/permissions.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (bsikander) [reply doc/permissions.md:1]: also, I think UI has a kill button which might need some thinking

[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply doc/permissions.md:1]: Ok, I will take a look.

File: job-server/src/main/resources/application.conf (hunk 16 above)
[the remark this answers is not in the document]
Comment (bsikander) [reply job-server/src/main/resources/application.conf:322]: Next release will either be 1.0.0 or 0.11.0. For now, you can create a file in [`notes`](https://github.com/spark-jobserver/spark-jobserver/tree/master/notes) folder with name 0.11.0 and put the info inside. During release, if we choose a different number, we will move the information.

File: scalastyle-config.xml (hunk 13 above)
[the remark this answers is not in the document]
Comment (bsikander) [reply scalastyle-config.xml:75]: No, i do not have a better idea. Disabling it is fine. If somehow finds a way, they can contribute.

File: job-server/src/main/scala/spark/jobserver/auth/ShiroAuthenticator.scala [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+import scala.concurrent.ExecutionContext
+
+/**
+  * Apache Shiro based authenticator for the Spark JobServer, the authenticator realm must be
+  * specified in the ini file for Shiro
+  */
+class ShiroAuthenticator(override protected val config: Config)
+                        (implicit ec: ExecutionContext, s: ActorSystem) extends SJSAuthenticator(config) {
+
+  val sManager = new IniSecurityManagerFactory(config.getString("shiro.config.path")).getInstance()
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/ShiroAuthenticator.scala:21]: The config for Shiro (and Keycloak) is always required if the according authentication provider is used. If the configuration is actually missing, it is a mis-configuration by the user. In that case a `ConfigException.Missing` exception is raised during startup. I don't think a graceful handling is a good idea as there is no way for us to recover from missing configs.

File: job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+import java.security.Key
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.Try
+
+class KeycloakAuthenticator(override protected val config: Config)
+                           (implicit ec: ExecutionContext, s: ActorSystem)
+  extends SJSAuthenticator(config) {
+
+  private val baseUrl = f"${config.getString("keycloak.authServerUrl").stripSuffix("/")}/realms/" +
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala:26]: See comment on ShiroAuthenticator above.

File: job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+    .setSigningKeyResolver(new SigningKeyResolver {
+      override def resolveSigningKey(header: JwsHeader[T] forSome {type T <: JwsHeader[T]},
+                                     claims: Claims): Key =
+        resolveSigningKey(header, "")
+
+      override def resolveSigningKey(header: JwsHeader[T] forSome {type T <: JwsHeader[T]},
+                                     plaintext: String): Key = {
+        Try(keyStore.get(header.getKeyId).getPublicKey).recoverWith {
+          case ex: JwkException =>
+            logger.warn(s"Failed to fetch JWK: ${ex.getMessage}")
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala:42]: This method is called each time a JWT is parsed. If fetching the public key fails, the verification of the token will also fail and consequently the user is not authenticated. This state does not have to be permanent, e.g. due to network issues. I raised the log level to `error`.

File: job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+            null
+        }.get
+      }
+    })
+
+  private val cache: Cache[String, AuthInfo] = LfuCache.apply
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala:48]: Default configuration is available at [akka-http-cache](https://github.com/akka/akka-http/blob/v10.2.2/akka-http-caching/src/main/resources/reference.conf). After having thought about it again, I think caching the futures indefinitely for less than 512 users is a very bad idea. I will have another look at it. Maybe it is reasonable to also provide caching for shiro...

File: job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+      uri = f"$baseUrl/token",
+      entity = FormData(payload).toEntity)
+    Http()
+      .singleRequest(request)
+      .map(res => JsonUtils.mapFromJson(
+        res.entity.asInstanceOf[HttpEntity.Strict].data.utf8String))
+      .map(m => {
+        m.get("access_token")
+          .map(_.asInstanceOf[String])
+          .getOrElse(throw new IncorrectCredentialsException)
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala:93]: There is no equivalent in Keycloak. But maybe it is better to introduce a custom exception that is used in all authenticators.

File: job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+      "username" -> username,
+      "password" -> password,
+      "client_id" -> clientId
+    ) ++ clientSecret.map("client_secret" -> _)
+
+    val request = HttpRequest(
+      method = HttpMethods.POST,
+      uri = f"$baseUrl/token",
+      entity = FormData(payload).toEntity)
+    Http()
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala:86]: As the provided credentials of the user can not be validated, the user will not be authenticated and all requests are canceled with `401 Unauthorized`. However, as the failed result is not cached, each subsequent request will try to reach Keycloak again. Only after a successful authentication, the result is cached.

File: job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+import spark.jobserver.common.akka.web.JsonUtils
+
+import scala.collection.JavaConverters._
+import java.net.URL
+import java.security.Key
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.Try
+
+class KeycloakAuthenticator(override protected val config: Config)
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala:22]: See comment on failure handling above.

File: job-server/src/main/scala/spark/jobserver/auth/ShiroAuthenticator.scala (hunk 17 above)
[the remark this answers is not in the document]
Comment (bsikander) [reply job-server/src/main/scala/spark/jobserver/auth/ShiroAuthenticator.scala]: Didn't notice that the missing exception is already being thrown. In that case, it seems fine.

**Comment by Ennosigaeon** [comment]
@bsikander: I have added three new commits containing partial solutions for you feedback. There are still some points I will have a more detailed look at in the next days. Regarding your questions:
1. We use spark-jobserver as a central component of our spark cluster. We want to limit *write* access to spark-jobserver to prevent accidental changes but still provide a *read* access for everybody to access job results. Similarly, we want to prevent people from accidentally overriding binaries run in production.
2. Maybe it is better to rename the config section to `authorization`. Or do you have any preferences?

File: README.md [hunk 24]
```diff
@@ -573,11 +589,41 @@ securityManager.cacheManager = $cacheManager
 
 Make sure to edit the url, credentials, userDnTemplate, ldap.allowedGroups and ldap.searchBase settings in accordance with your local setup.
 
-Here is an example of a simple curl command that authenticates a user and uses ssl (you may want to use -H to hide the
-credentials, this is just a simple example to get you started):
+#### Keycloak Authentication
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply README.md:594]: Ok, I had another look at Shiro and just implemented authorization management. However, I was only test it using the `dummy.ini` file as I don't have access to an active directory.

File: job-server/src/main/resources/application.conf (hunk 16 above)
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/resources/application.conf:322]: I have renamed `authentication` to `access-control` and added a 0.11.0 release notes.

File: job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala (hunk 21 above)
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala:93]: I have replaced the Shiro `IncorrectCredentialsException` with a custom implementation.

File: README.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply README.md:594]: And I have rephrased the commit message for the keyclaok feature to include some (minor) motivation.

File: job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala (hunk 20 above)
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala]: Ok, I have added caching for shiro as well. Configuration can now be explicitly provided via `application.conf`, reasonable default configuration is provided.

**Comment by Ennosigaeon** [comment]
@bsikander: I have to admit that I somewhat lost the overview which remarks of you are still relevant/open. Could you please resolve all issues that are done?

Open Issues:
- Test UI with limited permissions
- Check if retry handling should be added

**Comment by Ennosigaeon** [comment]
Ok, I have tested the UI. If a permission to read jobs, contexts or binaries is missing, the request to load the data silently fails and the resulting table is simply empty with an exception logged to console. I have added handling of missing permissions to kill a job.

File: README.md [hunk 25]
```diff
@@ -543,13 +546,28 @@ The minimum set of parameters to enable client authentication consists of:
...
 ```
-authentication = on
-# absolute path to shiro config file, including file name
-config.path = "/some/path/shiro.ini"
+
+#### Shiro Authentication
+The Shiro Authenticator can be activated in the configuration file by changing the authentication provider and
+providing a shiro configuration file.
+```
+authentication {
```
Comment (bsikander) [inline README.md:565]: nit:
access-control

File: README.md [hunk 26]
```diff
@@ -573,11 +591,46 @@ securityManager.cacheManager = $cacheManager
...
+The Shiro authenticator is able to perform fine-grained [user authorization](#user-authorization). Permissions are
+extracted from the provided user roles. Each role that matches a known permission is added to the authenticated user.
+Unknown roles are ignored. For a list of available permissions see [Permissions](doc/permissions.md).
+
+#### Keycloak Authentication
+The Keycloak Authenticator can be activated in the configuration file by changing the authentication provider and
+providing a keycloak configuration.
 ```
-curl -k --basic --user 'user:pw' https://localhost:8090/contexts
+authentication {
```
Comment (bsikander) [inline README.md:602]: nit:
access-control

File: doc/permissions.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (bsikander) [reply doc/permissions.md:1]: +1 for securing the UI also and mentioning this in the permissins.md

File: job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala (hunk 21 above)
[the remark this answers is not in the document]
Comment (bsikander) [reply job-server/src/main/scala/spark/jobserver/auth/KeycloakAuthenticator.scala]: It is fine that you have added the exception directly in the class but jobserver has a common place for all custom exceptions. https://github.com/Ennosigaeon/spark-jobserver/blob/master/job-server/src/main/scala/spark/jobserver/util/CustomException.scala

**Review by bsikander — COMMENTED** [review]
I have resolved all the comments, I will review the new changes that you have done and will also check the UI.

File: README.md (hunk 25 above)
Comment (Ennosigaeon) [reply README.md]: Ah, missed those.

File: job-server/src/main/resources/application.conf [hunk 27]
```diff
@@ -319,17 +319,33 @@ akka.http.server {
...
+  }
+
+  keycloak {
+    authServerUrl = "https://example.com/auth"
+    realmName = "realm"
+    client = "client"
+    clientSecret = "clientSecret"
+  }
+
+  akka.http.caching.lfu-cache {
```
Comment (bsikander) [inline job-server/src/main/resources/application.conf:345]: Are these settings automatically picked up even if they are inside `access-control` section?

Comment (Ennosigaeon) [reply job-server/src/main/resources/application.conf:345]: Yes, I only pass the `access-control` subsection to the cache settings creation.

**Comment by bsikander** [comment]
I did some testing, caching might not be the best idea atleast for basic authentication. If the first time, I pass in correct username and password, it caches it but afterwards, even with wrong password, I am able to do all operations. I tried this with shiro.

**Comment by Ennosigaeon** [comment]
The caching indeed stores the verification of the credentials. Once you have been authenticated, all following authentications are skipped for the configured cache time. 

Caching itself is a compromise between security and performance that could still be useful (see [Shiro Caching](https://shiro.apache.org/caching.html)). Maybe, I can make caching optional with disabled by default? So users can decide on their own.

I can remove Caching from Shiro completely as Shiro implements custom caching mechanisms (I just wanted Shiro and Keycloak to behave somewhat similar for simplicity). But for Keycloak, I can become quite expensive to perform multiple HTTP round trips to the Keycloak server for each individual request.

**Comment by bsikander** [comment]
I have not used shiro with LDAP and never tried its caching mechanism. So, I am finding it hard to suggest you a solution.

I liked the idea of both implementations having caching with an option to disable it through config parameter. This allows users to disable jobserver caching and use Shiro caching. This way user have both options. 

The above is just a suggestion, if you feel like we should leave this upon future contributors who has concrete requirements for Shiro caching, please go ahead and remove the caching from Shiro.

**Comment by bsikander** [comment]
Apart from above, somehow, authorization in shiro is not working.
Here is my shiro.ini
```
[users]
admin = test, admin
basic = test1, binOnly, jobOnly


[roles]
admin = *
binOnly = binaries:read
jobOnly  = jobs:read
```

When I try to access `/contexts` which should not be allowed, it returns me the output instead of 403.
`curl -u "basic:test1" localhost:8090/contexts`

**Comment by Ennosigaeon** [comment]
I have just checked for the direct assignment of roles, i.e.
```
[users]
admin = test, jobs:read
```
I have added an additional check for permissions and roles. Should work now as intended.

Caching is now optional (disabled by default) and can be enabled via `access-control.use-cache`

File: job-server/src/main/resources/application.conf [hunk 28]
```diff
@@ -319,17 +319,36 @@ akka.http.server {
...
+
+  keycloak {
+    authServerUrl = "https://example.com/auth"
+    realmName = "realm"
+    client = "client"
+    clientSecret = "clientSecret"
+  }
+
+  # If turned on, access-control for authorized users is cached to improve performance. While a user is cached, the
+  # provided password is *not* validated introducing a potential security thread.
```
Comment (bsikander) [inline job-server/src/main/resources/application.conf:346]: nit:
threat

**Review by bsikander — APPROVED** [review]
+1 for reducing the ttl.
Now shiro is behaving correctly, thank you for the changes.

You can add the following to `shiro.ini.basic.template` in config folder to add roles example
```
[users]
user1 = password1, admin
user2 = password2, contextsAdmin
user3 = password3, jobReader

[roles]
admin = *
contextsAdmin = contexts
jobReader = jobs:read
```

This change is ready to be merged now.

**Comment by Ennosigaeon** [comment]
I have adapted the Shiro template, fixed the typo and squashed all commits.

**Comment by bsikander** [comment]
Thank you for the contribution.

**Pull request merged**